File: services/api-geo/src/surveys_canon.rs
// Surveys canoniques (unifiés)
use crate::state::AppState;
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    Json,
};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Serialize, FromRow)]
pub struct SurveyCanon {
    pub id: Uuid,
    pub code: String,
    pub localite_canon: String,
    pub localite: Option<String>,
    pub adm3_id: Option<i32>,
    pub adm3_name: Option<String>,
    pub geom: Option<serde_json::Value>,
    pub location_mode: Option<String>,
    pub is_geocoded: bool,
    pub has_geom: bool,
    pub has_adm3: bool,
    pub date: Option<chrono::NaiveDate>,
    pub nb_sondages_source: i32,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Deserialize)]
pub struct SurveysQuery {
    pub limit: Option<i64>,
    pub offset: Option<i64>,
    pub search: Option<String>,
    pub missing: Option<String>, // "geom" | "adm3"
}

/// GET /surveys - Liste des sondages unifiés
pub async fn list_surveys(
    State(state): State<AppState>,
    Query(params): Query<SurveysQuery>,
) -> Result<Json<Vec<SurveyCanon>>, (StatusCode, String)> {
    let pool = &state.pool;
    let limit = params.limit.unwrap_or(100).min(500);
    let offset = params.offset.unwrap_or(0);

    // Build WHERE clause based on filters
    let mut where_clauses = vec![];

    if params.search.is_some() {
        where_clauses.push(
            "(atlas.norm(code) LIKE '%' || atlas.norm($1) || '%' OR \
             atlas.norm(COALESCE(localite, '')) LIKE '%' || atlas.norm($1) || '%' OR \
             atlas.norm(localite_canon) LIKE '%' || atlas.norm($1) || '%')",
        );
    }

    if let Some(ref missing) = params.missing {
        if missing == "geom" {
            where_clauses.push("geom IS NULL");
        } else if missing == "adm3" {
            where_clauses.push("adm3_id IS NULL");
        }
    }

    let where_sql = if where_clauses.is_empty() {
        String::from("TRUE")
    } else {
        where_clauses.join(" AND ")
    };

    let query = format!(
        r#"
        SELECT 
            id, code, localite_canon, localite, adm3_id, adm3_name,
            ST_AsGeoJSON(geom)::jsonb as geom,
            location_mode::text as location_mode, is_geocoded, has_geom, has_adm3,
            date, nb_sondages_source, created_at, updated_at
        FROM atlas.surveys
        WHERE {}
        ORDER BY created_at DESC
        LIMIT $2 OFFSET $3
        "#,
        where_sql
    );

    let rows = sqlx::query_as::<_, SurveyCanon>(&query)
        .bind(&params.search)
        .bind(limit)
        .bind(offset)
        .fetch_all(pool)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(rows))
}

/// GET /surveys/resolve?code=XXX - Résoudre un alias vers le survey canonique
pub async fn resolve_alias(
    State(state): State<AppState>,
    Query(params): Query<SurveysQuery>,
) -> Result<Json<SurveyCanon>, (StatusCode, String)> {
    let pool = &state.pool;

    let code = params.search.ok_or((
        StatusCode::BAD_REQUEST,
        "Missing 'search' parameter".to_string(),
    ))?;

    let row = sqlx::query_as::<_, SurveyCanon>(
        r#"
        SELECT 
            s.id, s.code, s.localite_canon, s.localite, s.adm3_id, s.adm3_name,
            ST_AsGeoJSON(s.geom)::jsonb as geom,
            s.location_mode::text as location_mode, s.is_geocoded, s.has_geom, s.has_adm3,
            s.date, s.nb_sondages_source, s.created_at, s.updated_at
        FROM atlas.survey_aliases a
        JOIN atlas.surveys s ON s.id = a.survey_id
        WHERE a.alias_code = $1
        LIMIT 1
        "#,
    )
    .bind(&code)
    .fetch_one(pool)
    .await
    .map_err(|e| match e {
        sqlx::Error::RowNotFound => (
            StatusCode::NOT_FOUND,
            format!("No survey found for alias '{}'", code),
        ),
        _ => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()),
    })?;

    Ok(Json(row))
}

/// GET /surveys/:id - Détails d'un sondage unifié
pub async fn get_survey(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<SurveyCanon>, (StatusCode, String)> {
    let pool = &state.pool;

    let row = sqlx::query_as::<_, SurveyCanon>(
        r#"
        SELECT 
            id, code, localite_canon, localite, adm3_id, adm3_name,
            ST_AsGeoJSON(geom)::jsonb as geom,
            location_mode::text as location_mode, is_geocoded, has_geom, has_adm3,
            date, nb_sondages_source, created_at, updated_at
        FROM atlas.surveys
        WHERE id = $1
        "#,
    )
    .bind(id)
    .fetch_one(pool)
    .await
    .map_err(|e| match e {
        sqlx::Error::RowNotFound => (StatusCode::NOT_FOUND, "Survey not found".to_string()),
        _ => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()),
    })?;

    Ok(Json(row))
}

#[derive(Debug, Serialize)]
pub struct SurveysStats {
    pub total: i64,
    pub geocoded: i64,
    pub with_geom: i64,
    pub with_adm3: i64,
}

/// GET /surveys/stats - Statistiques des sondages unifiés
pub async fn get_surveys_stats(
    State(state): State<AppState>,
) -> Result<Json<SurveysStats>, (StatusCode, String)> {
    let pool = &state.pool;

    let (total, geocoded, with_geom, with_adm3): (i64, i64, i64, i64) = sqlx::query_as(
        r#"
        SELECT 
            COUNT(*) as total,
            COUNT(*) FILTER (WHERE is_geocoded) as geocoded,
            COUNT(*) FILTER (WHERE geom IS NOT NULL) as with_geom,
            COUNT(*) FILTER (WHERE adm3_id IS NOT NULL) as with_adm3
        FROM atlas.surveys
        "#,
    )
    .fetch_one(pool)
    .await
    .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(SurveysStats {
        total,
        geocoded,
        with_geom,
        with_adm3,
    }))
}

// ============================================================================
// ENDPOINTS D'ÉCRITURE - Géocodage
// ============================================================================

#[derive(Debug, Deserialize)]
pub struct UpdateGeometryPayload {
    pub mode: String,                    // "exact" | "adm"
    pub geom: Option<serde_json::Value>, // GeoJSON Point si mode=exact
    pub adm3_id: Option<i32>,            // ADM3 ID si mode=adm
}

/// PATCH /surveys-canon/:id/geometry - Mettre à jour la géométrie
pub async fn update_geometry(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(payload): Json<UpdateGeometryPayload>,
) -> Result<Json<SurveyCanon>, (StatusCode, String)> {
    let pool = &state.pool;

    // Validation
    if payload.mode == "exact" && payload.geom.is_none() {
        return Err((
            StatusCode::BAD_REQUEST,
            "Mode 'exact' requires 'geom'".to_string(),
        ));
    }
    if payload.mode == "adm" && payload.adm3_id.is_none() {
        return Err((
            StatusCode::BAD_REQUEST,
            "Mode 'adm' requires 'adm3_id'".to_string(),
        ));
    }

    // Trouver le sondage source (premier de la liste des alias)
    let source_id: Uuid = sqlx::query_scalar("SELECT id FROM atlas.surveys WHERE id = $1")
        .bind(id)
        .fetch_optional(pool)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "Survey not found".to_string()))?;

    // Mettre à jour le sondage source dans la table sondages
    if payload.mode == "exact" {
        let geom_json = payload.geom.as_ref().unwrap();
        sqlx::query(
            r#"
            UPDATE sondages 
            SET geom = ST_SetSRID(ST_GeomFromGeoJSON($1), 4326),
                location_mode = 'exact',
                updated_at = NOW()
            WHERE id = $2
            "#,
        )
        .bind(geom_json)
        .bind(source_id)
        .execute(pool)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    } else if payload.mode == "adm" {
        sqlx::query(
            r#"
            UPDATE sondages 
            SET adm3_id = $1,
                location_mode = 'adm',
                updated_at = NOW()
            WHERE id = $2
            "#,
        )
        .bind(payload.adm3_id)
        .bind(source_id)
        .execute(pool)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    }

    // Rafraîchir les vues canoniques
    sqlx::query("SELECT atlas.refresh_surveys()")
        .execute(pool)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    // Retourner le survey mis à jour
    let updated = sqlx::query_as::<_, SurveyCanon>(
        r#"
        SELECT 
            id, code, localite_canon, localite, adm3_id, adm3_name,
            ST_AsGeoJSON(geom)::jsonb as geom,
            location_mode::text as location_mode, is_geocoded, has_geom, has_adm3,
            date, nb_sondages_source, created_at, updated_at
        FROM atlas.surveys
        WHERE id = $1
        "#,
    )
    .bind(id)
    .fetch_one(pool)
    .await
    .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(updated))
}

// ============================================================================
// CANDIDATS ADM3 - Suggestions basées sur similarité
// ============================================================================

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct Adm3Candidate {
    pub adm3_id: i32,
    pub gid: i32,
    pub name: String,
    pub code: Option<String>,
    pub adm2_name: Option<String>,
    pub score: f32, // similarity() returns FLOAT4
}

#[derive(Debug, Serialize)]
pub struct Adm3CandidatesResponse {
    pub survey_id: Uuid,
    pub localite: String,
    pub candidates: Vec<Adm3Candidate>,
}

/// GET /surveys-canon/:id/adm3-candidates - Obtenir les candidats ADM3 suggérés
pub async fn get_adm3_candidates(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<Adm3CandidatesResponse>, (StatusCode, String)> {
    let pool = &state.pool;

    // Récupérer le survey
    let survey: SurveyCanon = sqlx::query_as(
        r#"
        SELECT 
            id, code, localite_canon, localite, adm3_id, adm3_name,
            ST_AsGeoJSON(geom)::jsonb as geom,
            location_mode::text as location_mode, is_geocoded, has_geom, has_adm3,
            date, nb_sondages_source, created_at, updated_at
        FROM atlas.surveys
        WHERE id = $1
        "#,
    )
    .bind(id)
    .fetch_one(pool)
    .await
    .map_err(|e| match e {
        sqlx::Error::RowNotFound => (StatusCode::NOT_FOUND, "Survey not found".to_string()),
        _ => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()),
    })?;

    let search_term = survey
        .localite
        .as_ref()
        .unwrap_or(&survey.localite_canon);

    // Rechercher les candidats avec similarité
    let candidates: Vec<Adm3Candidate> = sqlx::query_as(
        r#"
        SELECT 
            gid as adm3_id,
            gid,
            adm3_fr as name,
            adm3_pcode as code,
            adm2_fr as adm2_name,
            similarity(atlas.norm(adm3_fr), atlas.norm($1)) as score
        FROM adm3
        WHERE similarity(atlas.norm(adm3_fr), atlas.norm($1)) > 0.3
        ORDER BY score DESC
        LIMIT 5
        "#,
    )
    .bind(search_term)
    .fetch_all(pool)
    .await
    .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(Adm3CandidatesResponse {
        survey_id: survey.id,
        localite: search_term.clone(),
        candidates,
    }))
}

File: ui/src/api/surveys-canon.ts
// API client pour surveys canoniques (unifiés)

export interface SurveyCanon {
  id: string;
  code: string;
  localite_canon: string;
  localite: string | null;
  localite_key: string | null;
  source: string | null;
  adm3_id: number | null;
  adm3_name: string | null;
  adm2_name: string | null;
  adm1_name: string | null;
  geom: Record<string, unknown> | null;
  geom_geojson: Record<string, unknown> | null;
  location_mode: string | null;
  location_accuracy: string | null;
  is_geocoded: boolean;
  has_geom: boolean;
  has_adm3: boolean;
  date: string | null;
  date_sondage: string | null;
  operator: string | null;
  notes: string | null;
  comment: string | null;
  meta: string | null;
  nb_sondages_source: number;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface SurveyCanonStats {
  total: number;
  geocoded: number;
  with_geom: number;
  with_adm3: number;
}

export interface SurveyCanonQuery {
  limit?: number;
  offset?: number;
  search?: string;
  missing?: 'geom' | 'adm3';
}

// ============================================================================
// BASE API - Utilise le module centralisé api-base
// ============================================================================

import { buildApiUrl as buildUrl, API_BASE } from '../api-base'

console.log('[surveys-canon] ✓ API_BASE:', API_BASE)

// Alias pour compatibilité interne
const buildApiUrl = buildUrl

export async function apiGet<T>(path: string, params?: Record<string, any>): Promise<T> {
  let url = buildApiUrl(path);
  
  // Ajouter les query params si présents
  if (params && Object.keys(params).length > 0) {
    const searchParams = new URLSearchParams();
    Object.entries(params).forEach(([k, v]) => {
      if (v !== undefined && v !== null) {
        searchParams.set(k, String(v));
      }
    });
    const queryString = searchParams.toString();
    if (queryString) {
      url += '?' + queryString;
    }
  }
  
  console.debug('[API] GET', url);
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`HTTP ${res.status} ${url}`);
  }
  return res.json() as Promise<T>;
}

export async function apiPatch<T>(path: string, body: any): Promise<T> {
  const urlString = buildApiUrl(path);
  const res = await fetch(urlString, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const error = await res.text();
    throw new Error(`HTTP ${res.status}: ${error}`);
  }
  return res.json() as Promise<T>;
}

export async function apiDelete(path: string): Promise<void> {
  const urlString = buildApiUrl(path);
  const res = await fetch(urlString, {
    method: 'DELETE',
  });
  if (!res.ok) {
    const error = await res.text();
    throw new Error(`HTTP ${res.status}: ${error}`);
  }
}

export async function listSurveysCanon(query?: SurveyCanonQuery): Promise<SurveyCanon[]> {
  return apiGet<SurveyCanon[]>('/surveys-canon', query);
}

export async function getSurveyCanon(id: string): Promise<SurveyCanon> {
  return apiGet<SurveyCanon>(`/surveys-canon/${id}`);
}

export async function resolveSurveyAlias(code: string): Promise<SurveyCanon> {
  return apiGet<SurveyCanon>('/surveys-canon/resolve', { search: code });
}

export async function getSurveyCanonStats(): Promise<SurveyCanonStats> {
  return apiGet<SurveyCanonStats>('/surveys-canon/stats');
}

// Helper pour déterminer le statut de localisation
export function getLocationStatus(survey: SurveyCanon): 'precise' | 'adm3' | 'missing' {
  if (survey.has_geom) return 'precise';
  if (survey.has_adm3) return 'adm3';
  return 'missing';
}

export function getLocationLabel(survey: SurveyCanon): string {
  const status = getLocationStatus(survey);
  switch (status) {
    case 'precise': return 'Précise';
    case 'adm3': return 'Par commune (ADM3)';
    case 'missing': return 'Manquante';
  }
}

// ============================================================================
// TYPES POUR GÉOCODAGE
// ============================================================================

export interface Adm3Candidate {
  adm3_id: number;
  gid: number;
  name: string;
  code: string | null;
  adm2_name: string | null;
  score: number;
}

export interface Adm3CandidatesResponse {
  survey_id: string;
  localite: string;
  candidates: Adm3Candidate[];
}

export interface UpdateGeometryPayload {
  mode: 'exact' | 'adm';
  geom?: { type: 'Point'; coordinates: [number, number] };
  adm3_id?: number;
}

// ============================================================================
// ENDPOINTS GÉOCODAGE
// ============================================================================

export async function getAdm3Candidates(surveyId: string): Promise<Adm3CandidatesResponse> {
  return apiGet<Adm3CandidatesResponse>(`/surveys-canon/${surveyId}/adm3-candidates`);
}

export async function updateSurveyGeometry(
  surveyId: string,
  payload: UpdateGeometryPayload
): Promise<SurveyCanon> {
  const urlString = buildApiUrl(`/surveys-canon/${surveyId}/geometry`);
  const response = await fetch(urlString, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`HTTP ${response.status}: ${error}`);
  }
  
  return response.json();
}
